File: src/main/java/com/handwoong/everyonewaiter/api/MemberApiController.java
package com.handwoong.everyonewaiter.api;

import com.handwoong.everyonewaiter.dto.BasicResponseDto;
import com.handwoong.everyonewaiter.dto.member.MemberPasswordDto;
import com.handwoong.everyonewaiter.service.MemberService;
import com.handwoong.everyonewaiter.utils.validation.DeleteValidationGroup;
import com.handwoong.everyonewaiter.utils.validation.UpdateValidationGroup;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@Slf4j
@RestController
@RequestMapping("/members")
@RequiredArgsConstructor
public class MemberApiController {

    private final MemberService memberService;

    @PutMapping("/edit/password")
    public ResponseEntity<BasicResponseDto> changePassword(
            Authentication authentication,
            @RequestBody @Validated(UpdateValidationGroup.class) MemberPasswordDto passwordDto) {
        String username = authentication.getName();
        log.info("[PUT] 회원 비밀번호 수정 = 로그인 아이디 : '{}'", username);
        BasicResponseDto body = memberService.changePassword(username, passwordDto);
        return new ResponseEntity<>(body, HttpStatus.OK);
    }

    @DeleteMapping("/leave")
    public ResponseEntity<BasicResponseDto> leaveMember(
            Authentication authentication,
            @RequestBody @Validated(DeleteValidationGroup.class) MemberPasswordDto passwordDto) {
        String username = authentication.getName();
        log.info("[DELETE] 회원 삭제 = 로그인 아이디 : '{}'", username);
        BasicResponseDto body = memberService.deleteMember(username, passwordDto);
        return new ResponseEntity<>(body, HttpStatus.OK);
    }
}

File: src/main/kotlin/com/handwoong/everyonewaiter/controller/member/MemberApiController.kt
package com.handwoong.everyonewaiter.controller.member

import com.handwoong.everyonewaiter.dto.member.MemberCreateRequest
import com.handwoong.everyonewaiter.dto.member.MemberLoginRequest
import com.handwoong.everyonewaiter.dto.member.TokenResponse
import com.handwoong.everyonewaiter.service.member.MemberService
import com.handwoong.everyonewaiter.util.createCookie
import org.springframework.http.HttpStatus.*
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import javax.servlet.http.HttpServletResponse
import javax.validation.Valid

@RestController
class MemberApiController(
    private val memberService: MemberService,
) {

    @PostMapping("/members/login")
    fun loginMember(
        response: HttpServletResponse,
        @RequestBody @Valid memberDto: MemberLoginRequest,
    ): TokenResponse {
        val tokenResponse = memberService.login(memberDto)
        response.addCookie(createCookie(tokenResponse.accessToken))
        return tokenResponse
    }

    @PostMapping("/members")
    fun registerMember(
        @RequestBody @Valid memberDto: MemberCreateRequest,
    ): ResponseEntity<Unit> {
        memberService.register(memberDto)
        return ResponseEntity<Unit>(CREATED)
    }

}
